src/tools/kdb/command.h

```diff
@@ -0,0 +1,204 @@
+	elektraFree (fmtBuffer);                                                                                                           \
+	return c;
+
+#define EXEC_EXT(prog, argv, status)                                                                                                       \
+	pid_t extPid;                                                                                                                      \
+	int timeout = 1000;                                                                                                                \
+	if (0 == (extPid = fork ()))                                                                                                       \
+	{                                                                                                                                  \
+		if (-1 == execve (prog, (char **) (argv), NULL))                                                                           \
+		{                                                                                                                          \
+			perror ("child process execve failed [%m]");                                                                       \
+			return -1;                                                                                                         \
+		}                                                                                                                          \
+	}                                                                                                                                  \
+	while (0 == waitpid (extPid, status, WNOHANG))                                                                                     \
+	{                                                                                                                                  \
+		if (--timeout < 0)                                                                                                         \
+		{                                                                                                                          \
+			perror ("timeout");                                                                                                \
+			return -1;                                                                                                         \
+		}                                                                                                                          \
+		sleep (1);                                                                                                                 \
+	}
+
+// only print if we are at least as 'verbose' as minLogLevel
+#define CLI_PRINT(minLogLevel, fmt, ...) cliPrint (logLevel, minLogLevel, fmt, __VA_ARGS__)
+#define CLI_ERROR_PRINT(minLogLevel, fmt, ...) cliErrorPrint (logLevel, minLogLevel, fmt, __VA_ARGS__)
```

Comment:
Not a fan of these macros, the only thing they avoid is having to pass `logLevel` explicitly. However, exactly that will cause weird errors, if the function is used where `logLevel` is undefined or defined differently.

Based on the implementation of `cliPrint` and `cliErrorPrint` it would actually make more sense to define the macros as:

```c
#define CLI_PRINT(minLogLevel, fmt, ...) if (logLevel < minLogLevel) { printf (fmt, __VA_ARGS__); }
#define CLI_ERROR_PRINT(minLogLevel, fmt, ...) if (logLevel < minLogLevel) { fprintf (stderr, fmt, __VA_ARGS__); }
```

Then you at least avoid the `va_list` stuff.

But I'd still prefer using the functions directly for a bit of type-safety

Reply:
All of these macros are supposed to be used in the context of a command, so after GET_BASIC_OPTIONS. The idea was that when implementing a command you just have to think about the command and what it should do, so you don't have to think about the context of a CLI.

Reply:
Hm okay, I get the idea now... Maybe you could do something like this:


In `command.h`:
```c
typedef struct {
   int logLevel;
   char * fmtBuffer;
   // ...
} CliContext;

void loadCliContext (CliContext * ctx, KeySet * options);

static inline getOptionKey (KeySet * options, const char * name)
{
    return ksLookupByName (options, name, 0);
}

void cliPrint (const CliContext * ctx, int minLogLevel, const char * fmt, ...);
// ...
```

In `comand.h` file:
```c
// might need errorKey argument, and error return code
void loadCliContext (struct CliContext * ctx, KeySet * options) {
   // not quite right, but you get the idea
   bool debug;
   elektraKeyToBoolean (GET_OPTION_KEY (options, "debug"), &debug);
   bool verbose;
   elektraKeyToBoolean (GET_OPTION_KEY (options, "verbose"), &verbose);
   ctx->logLevel = verbose + debug;
   ctx->fmtBuffer = elektraMalloc(1024);
}
```

And used like this:

```c
int execGet (KeySet * options, Key * errorKey)
{
  CliContext ctx;
  loadCliContext (&ctx, options);

  // ...

  cliPrint (&ctx, CLI_LOG_VERBOSE, "The resulting keyname is %s\n", keyName (found));
}
```

Then you avoid all the issues of complex macros and still get the same "I don't care how CLI options work" style of code in the actual command implementations.

But for this PR, just leave it as is. It's just an idea, you could look into for the next PR (if you find the time).

Reply:
I marked this as resolved, since the `struct` is a nice idea, but given the fact that the other PRs use the same structure and are already done, it's something for another day.